lib/player.js

```diff
@@ -3796,70 +3798,116 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
    * @param {string=} mimeType
    * @param {string=} codec
    * @param {string=} label
-   * @return {shaka.extern.Track}
+   * @return {!Promise.<shaka.extern.Track>}
    * @export
    */
-  addTextTrack(uri, language, kind, mimeType, codec, label) {
-    // TODO: Add an actual error for this.
+  async addTextTrack(uri, language, kind, mimeType, codec, label) {
+    if (this.loadMode_ != shaka.Player.LoadMode.MEDIA_SOURCE &&
+        this.loadMode_ != shaka.Player.LoadMode.SRC_EQUALS) {
+      shaka.log.error(
+          'Must call load() and wait for it to resolve before adding text ' +
+          'tracks.');
+      return Promise.reject(() => {
+        return new shaka.util.Error(
+            shaka.util.Error.Severity.RECOVERABLE,
+            shaka.util.Error.Category.PLAYER,
+            shaka.util.Error.Code.PLAYER_UNINITIALIZED);
+      });
+    }
+
+    if (!mimeType) {
+      // Try using the uri extension.
+      const extension = shaka.media.ManifestParser.getExtension(uri);
+      mimeType = {
+        'srt': 'text/srt',
+        'vtt': 'text/vtt',
+        'webvtt': 'text/vtt',
+        'ttml': 'application/ttml+xml',
+      }[extension];
+
+      if (!mimeType) {
+        try {
+          // eslint-disable-next-line require-atomic-updates
+          mimeType = await shaka.media.ManifestParser.getMimeType(uri,
+              this.createNetworkingEngine(),
+              this.config_.streaming.retryParameters);
+        } catch (error) {}
+      }
+
+      if (!mimeType) {
+        shaka.log.error(
+            'The mimeType has not been provided and it could not be deduced ' +
+            'from its extension.');
+        return Promise.reject(() => {
+          return new shaka.util.Error(
+              shaka.util.Error.Severity.RECOVERABLE,
+              shaka.util.Error.Category.TEXT,
+              shaka.util.Error.Code.TEXT_COULD_NOT_GUESS_MIME_TYPE,
+              extension);
+        });
+      }
+    }
+
     if (this.loadMode_ == shaka.Player.LoadMode.SRC_EQUALS) {
+      if (mimeType == 'text/srt') {
+        try {
+          const stringData = await this.getTextData_(uri,
+              this.createNetworkingEngine(),
+              this.config_.streaming.retryParameters);
+          const SrtTextParser = shaka.text.SrtTextParser;
```

Comment:
The only thing I don't like about this is that it makes the SRT text parser plugin required for all builds.  With this change, there would be no way to exclude it without failing the build.

We've never referred to an optional component in a core part of the library before, but it would be useful to be able to do so conditionally, so that features work when the plugin is there and don't when it's not.  I will experiment a bit with this to figure out the best way to make it work.

Reply:
I found something that works to solve this, but with an annoying caveat.

1. Remove `goog.require('shaka.text.SrtTextParser')`
1. Create a private method in `Player` with the annotation `@suppress {strictMissingProperties}`.  All access of `SrtTextParser` in `Player` happen in this method, so that the compiler won't complain if it's not in the build
1. In that private method, if `SrtTextParser` is not defined, throw

That will allow the library to be built and run either **with** or **without** the parser plugin.

**However**, here's the caveat:  it won't allow apps to remove the plugin from the build and add it later at runtime.  This shouldn't be a huge problem, but that would work for most other plugins, since they aren't accessed from the core library by name.  An app developer who has done this with some other plugin might expect it to work with this one, so this could _potentially_ surprise someone.

What happens in this case is that without the plugin at build time, the compiler would rename this core access to `SrtTextParser` to some minified nonsense, but never connect that minified name to an implementation.  The runtime-supplied version of the plugin at the app level would be uncompiled and would not provide those minified names, so the Player would behave as if the plugin didn't exist at the `Player` level.  The plugin's registration with `TextEngine` would still be functional, so `TextEngine` could still find and execute the plugin for SRT content, but the static call to support `text/srt` in src= mode would not work.

So the problems with this solution are all fairly minor, and I would still be willing to accept the tradeoff.

Reply:
I have made the changes, but with the new compiler from a few days ago I always have this error:

```
ERROR - [JSC_MISSING_REQUIRE_IN_PROVIDES_FILE] 'shaka.text.SrtTextParser' references a namespace which was not required by this file.
Please add a goog.require.
  3973|         return shaka.text.SrtTextParser.srt2webvtt(data);
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 error(s), 0 warning(s), 94.5% typed
[ERROR] Build failed
```

Can you tell me what I'm doing wrong?